upstream/mos_prediction/expert.py

```diff
@@ -0,0 +1,104 @@
+from collections import OrderedDict
+from typing import List, Union, Dict
+
+import torch
+import torch.nn as nn
+from torch.tensor import Tensor
+from torch.nn.utils.rnn import pad_sequence
+from torchaudio.sox_effects import apply_effects_tensor
+from itertools import accumulate
+
+from upstream.interfaces import UpstreamBase, Featurizer
+from .model import MosDownstream, MosDownstreamModule
+from .utility import unfold_segments
+
+
+class UpstreamExpert(UpstreamBase):
+    def __init__(self, ckpt: str = None, model_config: str = None, **kwargs):
+        """
+        Args:
+            ckpt:
+                The checkpoint path for loading your pretrained weights.
+                Can be assigned by the -k option in run_downstream.py
+
+            model_config:
+                The config path for constructing your model.
+                Might not needed if you also save that in your checkpoint file.
+                Can be assigned by the -g option in run_downstream.py
+        """
+        # Pass kwargs into UpstreamBase to enable features shared across upstreams
+        super().__init__(**kwargs)
+
+        # The model needs to be a nn.Module for finetuning, not required for representation extraction
+        self.checkpoint = torch.load(ckpt, map_location="cpu")
+
+        self.upstream_type = kwargs['upstream']
+
+        self.mos_upstream = self._get_mos_upstream()
+        self.mos_featurizer = self._get_mos_featurizer()
+        self.mos_downstream = self._get_mos_downstream()
+
+        self.segments_durations = 1
+
+
+    def forward(
+        self, wavs: List[Tensor]
+    ) -> Dict[str, Union[Tensor, List[Tensor], Dict[str, Tensor]]]:
+        """
+        When the returning Dict contains the List or Dict with more than one Tensor,
+        those Tensors should be in the same shape if one wished to weighted sum them.
+        """
+        wavs_segments = [self.preprocessor(wav) for wav in wavs]
+
+        flattened_wavs_segments = [
+            wav_segment
+            for wav_segments in wavs_segments
+            for wav_segment in wav_segments
+        ]
+        wav_segments_lengths = [len(wav_segments) for wav_segments in wavs_segments]
+        prefix_sums = list(accumulate(wav_segments_lengths, initial=0))
+
+        
+        features = self.mos_upstream(flattened_wavs_segments)
```

Comment:
You might need to update the upstream usage.
Originally we input a list of waveform and output a list of representation.
Now we change to output a dictionary, whose `last_hidden_state` key correspond to the original representation you used.
From this point of view, I think the current state of this PR will not work on the latest master. You might need to first merge the master into your current branch and test it work.